BorderBoxControl: Refactor stories to TypeScript and Controls

aaronrobertshaw opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/35744

## What?
Finishes off conversion of BorderBoxControl to TypeScript, including updating stories to use controls instead of knobs.

## Why?
- To autogenerate docs in Storybook.

## How?
- Update export of `BorderBoxControl` named component so the type data can be extracted.
- Marks `BorderBoxControl` as non-polymorphic.
- Adds usage info and example from README as main JSDoc for the component.
- Converts stories to TypeScript.
- Removed the "default border" value as it added unwarranted complexity with the current approach to these stories.

## Testing Instructions

1. `npm run storybook:dev`
2. Check the [`BorderBoxControl` story](http://localhost:50240/?path=/story/components-experimental-borderboxcontrol--default) still function as expected
3. Check the generated docs.

aaronrobertshaw committed:
BorderBoxControl: Refactor storybook to TypeScript

aaronrobertshaw committed:
Add changelog

ciampo commented on packages/components/src/border-box-control/stories/index.tsx:
```diff
@@ -35,72 +52,57 @@ const colors = [
-export const Default = _default.bind( {} );
+export const Default = Template.bind( {} );
 Default.args = {
+	colors,
```
The "default" story should have the least amount of args possible, since it should showcase the default behaviour of the component — in this specific case, it looks like we could probably just leave `label` and `colors` and remove the rest.

This approach can also help us to understand if the component has sensible defaults, or if we need to make tweaks (as written above).

If we wanted to highlight a particular scenario (e.g allowing transparency, or disabling border styles), we could create a specific story for it

<details>

<summary>For example</summary>

```diff
diff --git a/packages/components/src/border-box-control/stories/index.tsx b/packages/components/src/border-box-control/stories/index.tsx
index 1717db3281..cebed8b2bc 100644
--- a/packages/components/src/border-box-control/stories/index.tsx
+++ b/packages/components/src/border-box-control/stories/index.tsx
@@ -100,9 +100,25 @@ export const Default = Template.bind( {} );
 Default.args = {
 	colors,
 	label: 'Borders',
-	disableCustomColors: false,
+};
+
+export const EnableAlpha = Template.bind( {} );
+EnableAlpha.args = {
+	...Default.args,
+	label: 'Enable alpha',
 	enableAlpha: true,
-	enableStyle: true,
-	size: 'default',
-	popoverPlacement: 'right-start',
+};
+
+export const DisableBorderStyles = Template.bind( {} );
+DisableBorderStyles.args = {
+	...Default.args,
+	label: 'Disable border styles',
+	enableStyle: false,
+};
+
+export const AllowCustomColors = Template.bind( {} );
+AllowCustomColors.args = {
+	...Default.args,
+	label: 'Allow custom colors',
+	disableCustomColors: false,
 };
```

</details>

aaronrobertshaw replied:
Thanks, I don't think those extra defaults were really required and don't warrant their own standalone stories given they can just be toggled on/off in the controls easy enough.

I've removed them https://github.com/WordPress/gutenberg/pull/45002/commits/364c803feb1ba8060c61e47903118deed753d95a

ciampo replied:
Agreed

ciampo commented on packages/components/src/border-box-control/stories/index.tsx:
```diff
@@ -35,72 +52,57 @@ const colors = [
-export const Default = _default.bind( {} );
+export const Default = Template.bind( {} );
 Default.args = {
+	colors,
```
Nit: Let's make the `colors` array shorter, so that the code example generated in the docs tab remains a bit more compact

aaronrobertshaw replied:
I've reduced the available colors as well as fixed up the duplicates that were in the `BorderControl` story after its recent refactor.

https://github.com/WordPress/gutenberg/pull/45002/commits/89cf202f3f00f47d94f3db8e850bc234743721d3

ciampo commented on packages/components/src/border-box-control/stories/index.tsx:
```diff
@@ -35,72 +52,57 @@ const colors = [
+			>
 				Reset
 			</Button>
+			{ /* @ts-expect-error Ignore until Popover is converted to TS */ }
```
Nit: Popover is actually converted to TypeScript, but its slot is not
 
```suggestion
			{ /* @ts-expect-error Ignore until Popover.Slot gets typed for the Popover component */ }
```

aaronrobertshaw replied:
Thanks. It's good to be accurate. I've fixed this here and in the `BorderControl` where I copied it from 🙂 

https://github.com/WordPress/gutenberg/pull/45002/commits/8277b809a6475513c602d1502f3a1062f5667a84

ciampo commented on packages/components/src/border-box-control/stories/index.tsx:
```diff
@@ -35,72 +52,57 @@ const colors = [
-};
+const Template: ComponentStory< typeof BorderBoxControl > = ( props ) => {
+	const { onChange, ...otherProps } = props;
+	const [ borders, setBorders ] = useState< AnyBorder >();
```
No need to use the `AnyBorder` type, we can use the `value` prop type instead:

```suggestion
	const [ borders, setBorders ] = useState< typeof props[ 'value' ] >();
```

aaronrobertshaw replied:
✅ Done:  https://github.com/WordPress/gutenberg/pull/45002/commits/2edad2ea69b5c6ca137ff5eeff094ae9bb1365d9

ciampo reviewed (commented):
Thank you Aaron, this is definitely an improvement over the current Storybook examples, and a great way to have more insights on this component's docs and APIs.

---

In fact, as I reviewed this PR, I noticed a few inconsistencies that should be probably be ironed out:

- most props lack a default value. While this is not a mandatory, I think it would make the docs much more clear and set better expectations on what is the component's default behavior (i.e. are custom colors enabled by default? are border styles enabled by default? etc.)
- I actually noticed a couple of inconsistencies with the docs:
  - the `enableStyle` prop should show a default value of `true`, since it's forwarded to `BorderControl` components where that prop has a default value of `true`
  - the `colors` prop is marked as not required, but in my tests the component breaks when it's omitted (the error actually happens in the `ColorPalette` component) — an easy fix could be to assign a default value of `[]`
- especially for props that are simply forwarded to other components (like `BorderControl` and `Popover`), we should try to reuse their types (ie. inherit / pick their props when possible)
- there could be other inconsistencies that I didn't spot — which highlights how valuable it can be to write Storybook examples and unit tests in TypeScript

I'm ok with fixing these inconsistencies in a follow-up PR, so that we can focus on merging the few PRs on `BorderBoxControl` that are currently opened.

aaronrobertshaw committed:
Shorten available colors array

Also fixes the duplicate colors that were in the BorderControl's shortened color array.

aaronrobertshaw committed:
Improve accuracy of ts expect error comment

aaronrobertshaw committed:
Clean up border values type in story

aaronrobertshaw committed:
Remove extraneous default args from main story

aaronrobertshaw commented:
Appreciate the review and detailed explanations @ciampo 👍 

As you suggest, I'll create a separate PR to iron out any inconsistencies with prop defaults, types, and documentation. 

In the meantime, the other feedback for this PR has been addressed so 🤞 it should be right to land when I get a final approval on https://github.com/WordPress/gutenberg/pull/41860.

ciampo reviewed (approved):
Sounds like a plan 🚀

ciampo commented:
FYI - `ColorPalette` is being typed in #44632 — the follow-up work refining `BorderBoxControl`'s types may benefit from reusing some of the types being introduced there :)

ciampo commented:
Update — #44632 has been merged (in case you were going to open that follow-up on ironing out `BorderBoxControl`'s type inconsistencies)
